text_extensions_for_pandas/cleaning/preprocess.py

```diff
@@ -0,0 +1,410 @@
+#  See the License for the specific language governing permissions and
+#  limitations under the License.
+#
+###############################################################################
+# preprocess.py
+#
+# Cleaning utilities for finding errors in varied corpora
+#
+
+import numpy as np
+import pandas as pd
+import transformers
+from transformers.file_utils import PANDAS_IMPORT_ERROR
+
+import text_extensions_for_pandas as tp
+from typing import *
+
+# Always run with the latest version of Text Extensions for Pandas
+import importlib
+
+tp = importlib.reload(tp)
+
+
+def preprocess_doc_with_bert(
+    document: pd.DataFrame,
+    bert_model,
+    tokenizer,
+    token_col,
+    label_col,
+    keep_cols: List[str],
+    iob: bool,
+    iob_col=None,
+):
+    """
+    Translates a single document in the form of a pandas DataFrame from standard spans into BERT compatible spans,
+    and calculates BERT embeddings over that text, carries over iob spans properly, as well as any additional
+    information specified, on a token-by-token basis, and ensures that iob-spans are treated properly
+    :param document: a PandasDataframe as produced by :func:`tp.io.conll.conll_2003_to_documents`.
+     or :func:`tp.io.conll_u_to_documents`. Must contain a column containing `span` elements and some form of
+     label column, or two if IOB format is being used.
+    :param bert_model: PyTorch-based BERT model from the `transformers` library.
+                       A default model will be used if this is `None` or not specified
+
+    :param tokenizer: A tokenizer that is a subclass of huggingface transformers
+                       PreTrainingTokenizerFast which supports `encode_plus` with
+                       return_offsets_mapping=True.
+                       A default tokenizer will be used if this is `None` or not specified
+    :param token_col: the column in the each of the dataframes in `doc` containing the spans
+     for each individual token
+    :param label_col: the name of the pandas column `document` containing the label
+     over which you wish to classify (or identify incorrect elements). If using iob format
+     this should be the entity type label, not the IOB label
+    :param iob_col: if in iob format this must be specified. The column containing the IOB
+     tag portion of the label. If using output from :func:`tp.io.conll.conll_2003_to_documents`
+     this should be `'ent_col'`
+    :param keep_cols: a list of names of columns that you desire to carry over to the new tokeniztion
+    :param iob: if true, additional logic will be done to ensure that iob entities are treated properly
+    """
+    # make bert tokens
+    bert_toks = tp.io.bert.make_bert_tokens(
+        document.loc[0, token_col].target_text, tokenizer
+    )
+    raw_tok_spans = (
+        tp.TokenSpanArray.align_to_tokens(bert_toks["span"], document[token_col])
+        .as_frame()
```

Comment:
Might want to skip the round-trip through a new DataFrame and just pull out the `begin_token` and `end_token` attributes of the return value from `align_to_tokens`

Reply:
Updated in latest commit